File: src/infrastructure/servers/factories/groq_server_factory.py
from openai import OpenAI
from infrastructure.servers.factories.abstracts.base_server_factory import BaseServerFactory

class GroqServerFactory(BaseServerFactory):
    def build_server(self) -> OpenAI:
        api_key: str = self.require_env(key="LLM__GROQ_API_KEY", error_message="Groq API key não encontrada no .env")
        base_url: str = self.optional_env(key="LLM__GROQ_BASE_URL", default="https://api.groq.com/openai/v1")
        timeout: float = self.optional_float_env(key="LLM__GROQ_TIMEOUT_SECONDS", default=30.0)
        max_retries: int = self.optional_int_env(key="LLM__GROQ_MAX_RETRIES", default=1)
        return OpenAI(
            api_key=api_key,
            base_url=base_url,
            timeout=timeout,
            max_retries=max_retries,
        )

File: angular-app/src/app/core/infrastructure/servers/factories/groq-server-factory.ts
import OpenAI from 'openai';
import type { IServer, ModelDescriptor, TextCompletionResponse } from '../abstracts/i-server';
import { BaseServerFactory } from './abstracts/base-server-factory';

export class GroqServerFactory extends BaseServerFactory {
  private readonly fallbackModels: ModelDescriptor[] = [
    { id: 'llama-3.1-8b-instant' },
  ];

  buildServer(): IServer {
    const apiKey = this.requireEnv('LLM__GROQ_API_KEY', 'Groq API key não encontrada no .env');
    const baseURL = this.optionalEnv('LLM__GROQ_BASE_URL', 'https://api.groq.com/openai/v1');
    const timeoutSeconds = this.optionalFloatEnv('LLM__GROQ_TIMEOUT_SECONDS', 30.0);
    const maxRetries = this.optionalIntEnv('LLM__GROQ_MAX_RETRIES', 1);
    const client = new OpenAI({
      apiKey,
      baseURL,
      timeout: timeoutSeconds * 1000,
      maxRetries,
    });

    return {
      chat: {
        completions: {
          create: async (params): Promise<TextCompletionResponse> => client.chat.completions.create(params),
        },
      },
      models: {
        list: async () => {
          try {
            const page = await client.models.list();
            return {
              data: page.data.map((model) => ({ id: model.id })),
            };
          } catch {
            return { data: this.fallbackModels };
          }
        },
      },
    };
  }
}
